doc/generated/man/irqbench.1
'\" t
.\"     Title: irqbench
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 2008/04/19
.\"    Manual: Xenomai Manual
.\"    Source: Xenomai 2.6.5
.\"  Language: English
.\"
.TH "IRQBENCH" "1" "2008/04/19" "Xenomai 2\&.6\&.5" "Xenomai Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
irqbench \- Xenomai IRQ benchmark, host control
.SH "SYNOPSIS"
.sp
\fBirqbench\fR [ options ]
.SH "DESCRIPTION"
.sp
\fBirqbench\fR is part of the Xenomai test suite and benchmarks IRQs\&. It is the controlling part running on a plain Linux host, connected via RS232 or Laplink with the target which is running a suitable Xenomai enabled kernel with the respective module (xeno_posix)\&. The target must run \fBirqloop(1)\fR\&.
.SH "OPTIONS"
.sp
\fBirqbench\fR accepts the following options:
.PP
\fB\-p <period_us>\fR
.RS 4
signal period, default=100 us
.RE
.PP
\fB\-T <test_duration_seconds>\fR
.RS 4
default=0, so ^C to end
.RE
.PP
\fB\-o <port_type>\fR
.RS 4
0=serial (default), 1=parallel
.RE
.PP
\fB\-a <port_io_address>\fR
.RS 4
default=0x3f8/0x378
.RE
.PP
\fB\-f\fR
.RS 4
freeze trace for each new max latency
.RE
.SH "SEE ALSO"
.sp
\fB/usr/share/doc/xenomai\-doc/txt/irqbench\&.txt, \fR\fBirqloop(1)\fR\&.
.SH "AUTHOR"
.sp
\fBirqbench\fR was written by Jan Kiszka\&. This man page was written by Roland Stigge\&.


ChangeLog
2009-03-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vxworks/semLib.c (sem_destroy_internal) : Add missing
	rescheduling. Fix sem_read_proc().

2009-03-19  Philippe Gerum  <rpm@xenomai.org>

	* include/asm-generic/wrappers.h, src/nucleus/shadow.c: Fix
	generic bits for 2.6.29.

2009-03-18  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/pipe.c: Fix streaming mode.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/nucleus/shadow.c: Trace syscall exits.

	* patching file include/asm-generic/syscall.h,
	ksrc/nucleus/shadow.c: Add xn_sys_current_info.

	* src/rtdk/rt_print.c: suspend/resume printer thread.

	* ksrc/skins/native/syscall.c, src/skins/native/task.c: Support
	T_WARNSW on task creation.

	* src/skins/native/task.c: Avoid lazy non-TLS rt_task_self.

	* src/skins/native/task.c: Silent migration before self-deletion.

	* src/skins, ksrc/nucleus/shadow.c: Rework initial Linux
	schedparam setting.

	* ksrc/nucleus, include/asm-generic, ksrc/arch/generic: Exclude
	CPUs from Xenomai use.

	* ksrc/nucleus/shadow.c: Cleanup taskexit hook.

	* ksrc/nucleus/shadow.c: Suppress switch warning upon signal
	receipt when the process is being debugged.

	* ksrc/nucleus/sched.c: Allow to change the watchdog timeout via a
	module parameter.

	* ksrc/arch/x86/usercopy_{32, 64}.c: Fix miscompilation of
	__do_strncpy_from_user.

	* src/rtdk, include/rtdk.h: Add assert_context helpers.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/skins/posix/syscall.c: Do not return -EFAULT on zero-length
	names. Instead, return -EINVAL when trying to create objects with
	empty names.

2009-03-13  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/mx.c: Streamline mutex support.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/task.c, src/skins/vrtx/task.c: Fix task creation.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/shadow.c, ksrc/nucleus/synch.c: Introduce debug
	option to detect mutexes held in relaxed sections.

2009-03-07  Philippe Gerum  <rpm@xenomai.org>

	* src/skins/posix/shm.c (__wrap_munmap): Forward to glibc upon EBADF.

	* ksrc/nucleus/Config.in, ksrc/arch/{powerpc, x86}/Config.in: Add
	missing SYS_STACKPOOLSZ switch.

2009-02-26  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/powerpc/patches: Update Adeos support to
	2.6.24.6-powerpc-DENX-2.4-03 (compatible with MPC5121-ADS
	hardware, see ksrc/arch/powerpc/patches/README).

2009-02-25  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/nmi.c: Rework NMI watchdog support.
	
2008-02-24  Stephane Fillod  <stephane.fillod@thomson.net>

	* src/skins/posix/shm.c (__wrap_mmap64): Make real call depend on
	configure detection (HAVE_MMAP64).
	
2009-02-24  Yi Li <liyi.dev@gmail.com>

	* configure.in: Remove -mfdpic switch. Set by default by
	bfin-linux-uclibc-gcc, and may cause error with bfin-uclinux-gcc.

2009-02-24  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/{arm, blackfin, powerpc}/hal.c (rthal_timer_request):
	Port x86 fix of race window.

2009-02-24  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/hal-common.c (rthal_timer_request): Fix race.

	* ksrc/nucleus/shadow.c (gatekeeper_thread): Do not refer to
	uninit sched slots.

2009-02-23  Paul Corner <paul_c@tuxcnc.org>

	* debian/prepare-patch.sh: Fix patch generation rules.

2009-02-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.19-x86-2.2-06,
	2.6.28.7-x86-2.2-06.

2009-02-17  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-05.

	* ksrc/arch/powerpc/patches: Upgrade to
	2.6.28.5-powerpc-DENX-2.4-04.

2009-02-13  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* src/skins/posix/shm.c (__wrap_mmap, __wrap_mmap64): Map only
	part of the underlying nucleus heap.

	* ksrc/nucleus/heap.c (xnheap_mmap): Allow partial mapping in
	user-space of a nucleus heap.

	* ksrc/nucleus/sched.c, include/nucleus/sched.h: Introduce
	sched->cpu field member.

2009-02-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pod.c (xnpod_start_thread): Run start hook for
	user-space threads as well.

2009-02-11  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/posix/shm.c (ftruncate): Remove GFP_USER hint.

2009-02-07  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/blackfin, include/asm-blackfin: Prepare for 2.6.28.

	* ksrc/arch/blackfin/patches: Upgrade to 2.6.28-1.9-00 (GIT).

	* ksrc/nucleus, include/nucleus: Introduce host tick deferral.

2009-02-02  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/syscall.c (rt_queue_write): Fix memory leak.

2009-02-01  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/heap, include/asm-generic/system.h, ksrc/skins: Fix
	auto-cleanup for shared heaps.

2009-01-31  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/drivers/testing/switchtest.c,
	  src/testsuite/switchtest/switchtest.c: Add stress mode to
	switchtest.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/ia64, include/asm-ia64: Itanium port discontinued.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-04,
	2.6.28.2-x86-2.2-05.

2009-01-27  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* include/asm-x86/bits/pod_{32, 64}.h (xnarch_leave_root): Fix FPU
	backup clobbering.
	
2009-01-27  Philippe Gerum  <rpm@eclipse.xenomai.org>

	* ksrc/nucleus, include/nucleus, ksrc/skins/posix, include/posix,
	sim, include/asm-sim: Introduce POSIX sporadic scheduling (PSS)
	class.

2009-01-21  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/arch/arm/patches: Upgrade to 2.6.28-arm-1.2-00.
	
2009-01-19  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pipe.c, ksrc/skins/native/pipe.c,
	ksrc/skins/rtai/fifo.c: Refactor callback management, fixing a
	severe latency issue, and a potential memory leak in both
	disconnect/release handlers.

	* ksrc/nucleus/pipe.c (xnpipe_disconnect, xnpipe_release):

src/testsuite/irqbench/irqbench.c
/*
 * Copyright (C) 2006 Jan Kiszka <jan.kiszka@web.de>.
 *
 * Xenomai is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Xenomai is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Xenomai; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>
#include <sys/io.h>
#include <sys/mman.h>


#define SERPORT                 0
#define PARPORT                 1

/* --- Serial port --- */

#define MCR_DTR                 0x01
#define MCR_RTS                 0x02
#define MCR_OUT2                0x08

#define MSR_DELTA               0x0F

#define LCR(base) (base + 3) /* Line Control Register */
#define MCR(base) (base + 4) /* Modem Control Register */
#define LSR(base) (base + 5) /* Line Status Register */
#define MSR(base) (base + 6) /* Modem Status Register */

/* --- Parallel port --- */

#define CTRL_INIT               0x04
#define CTRL_STROBE             0x10

#define DATA(base) (base + 0) /* Data register */
#define STAT(base) (base + 1) /* Status register */
#define CTRL(base) (base + 2) /* Control register */

static double tsc2ns_scale;
static long long min_lat = LONG_MAX;
static long long max_lat = LONG_MIN;
static long long avg_lat;
static long outer_loops;
static int warmup = 1;
static int terminate;
static unsigned long port_ioaddr = 0x3F8;
static int port_type = SERPORT;
static unsigned int toggle;
static long loop_avg;
static long long period = 100000;
static int trigger_trace;

static inline long long rdtsc(void)
{
	unsigned long long tsc;

	__asm__ __volatile__("rdtsc" : "=A" (tsc));
	return tsc;
}

static long tsc2ns(long long tsc)
{
	if (tsc > LONG_MAX || tsc < LONG_MIN) {
		fprintf(stderr, "irqbench: overflow (%lld ns)!\n",
			(long long)(tsc2ns_scale * (double)tsc));
		exit(2);
	}
	return (long)(tsc2ns_scale * (double)tsc);
}

static inline long long ns2tsc(long long ns)
{
	return (long long)(((double)ns) / tsc2ns_scale);
}

static void calibrate_tsc(void)
{
	FILE *proc;
	char *lineptr = NULL;
	size_t len;
	double cpu_mhz;

	proc = fopen("/proc/cpuinfo", "r");
	if (proc == NULL) {
		perror("irqbench: Unable to open /proc/cpuinfo");
		exit(1);
	}

	while (getline(&lineptr, &len, proc) != -1)
		if (strncmp(lineptr, "cpu MHz", 7) == 0) {
			sscanf(strchr(lineptr, ':') + 1, "%lf", &cpu_mhz);
			break;
		}

	if (lineptr)
		free(lineptr);
	fclose(proc);

	printf("CPU frequency: %.3lf MHz\n", cpu_mhz);

	tsc2ns_scale = 1000.0 / cpu_mhz;
}

static void sighand(int signal)
{
	if (warmup)
		exit(0);
	else
		terminate = 1;
}

static void do_inner_loop(void)
{
	long long start, delay, timeout;
	long lat;

	__asm__ __volatile__("cli");

	if (port_type ==  SERPORT) {
		start = rdtsc();

		toggle ^= MCR_RTS;
		outb(toggle, MCR(port_ioaddr));

		timeout = start + period * 100;
		while ((inb(MSR(port_ioaddr)) & MSR_DELTA) == 0 &&
		       rdtsc() < timeout);

		delay = rdtsc() - start;
	} else {
		int status = inb(STAT(port_ioaddr));

		outb(0x08, DATA(port_ioaddr));

		start = rdtsc();

		outb(0x00, DATA(port_ioaddr));

		timeout = start + period * 100;
		while (inb(STAT(port_ioaddr)) == status &&
		       rdtsc() < timeout);

		delay = rdtsc() - start;
	}

	if (!warmup) {
		lat = tsc2ns(delay);

		loop_avg += lat;
		if (lat < min_lat)
			min_lat = lat;
		if (lat > max_lat) {
			max_lat = lat;
			if (trigger_trace) {
				if (port_type == SERPORT) {
					toggle ^= MCR_DTR;
					outb(toggle, MCR(port_ioaddr));
				} else {
					outb(0x18, DATA(port_ioaddr));
					outb(0x10, DATA(port_ioaddr));
				}
			}
		}
	}

	__asm__ __volatile__("sti");

	while (rdtsc() < start + period);
}

int main(int argc, char *argv[])
{
	int ioaddr_set = 0;
	unsigned long long count = 1;
	int c;

	signal(SIGINT, sighand);
	signal(SIGTERM, sighand);
	signal(SIGHUP, sighand);
	signal(SIGALRM, sighand);

	calibrate_tsc();

	while ((c = getopt(argc, argv, "p:T:o:a:f")) != EOF)
		switch (c) {
		case 'p':
			period = atoi(optarg) * 1000;
			break;

		case 'T':
			alarm(atoi(optarg));
			break;

		case 'o':
			port_type = atoi(optarg);
			break;

		case 'a':
			port_ioaddr = strtol(optarg, NULL, 0);
			ioaddr_set = 1;
			break;

		case 'f':
			trigger_trace = 1;
			break;

		default:
			fprintf(stderr, "usage: irqbench [options]\n"
				"  [-p <period_us>]             # signal period, default=100 us\n"
				"  [-T <test_duration_seconds>] # default=0, so ^C to end\n"
				"  [-o <port_type>]             # 0=serial (default), 1=parallel\n"
				"  [-a <port_io_address>]       # default=0x3f8/0x378\n"
				"  [-f]                         # freeze trace for each new max latency\n");
			exit(2);
		}

	/* set defaults for parallel port */
	if (port_type == 1 && !ioaddr_set)
		port_ioaddr = 0x378;

	if (iopl(3) < 0) {
		fprintf(stderr, "irqbench: superuser permissions required\n");
		exit(1);
	}
	mlockall(MCL_CURRENT | MCL_FUTURE);

	switch (port_type) {
	case SERPORT:
		toggle = MCR_OUT2;
		inb(MSR(port_ioaddr));
		break;

	case PARPORT:
		outb(CTRL_INIT, CTRL(port_ioaddr));
		break;

	default:
		fprintf(stderr, "irqbench: invalid port type\n");
		exit(1);
	}

	period = ns2tsc(period);

	printf("Port type:     %s\n"
	       "Port address:  0x%lx\n\n",
	       (port_type == SERPORT) ? "serial" : "parallel", port_ioaddr);

	printf("Waiting on target...\n");

	while (1)
		if (port_type ==  SERPORT) {
			toggle ^= MCR_RTS;
			outb(toggle, MCR(port_ioaddr));
			usleep(100000);
			if ((inb(MSR(port_ioaddr)) & MSR_DELTA) != 0)
				break;
		} else {
			int status = inb(STAT(port_ioaddr));

			outb(0x08, DATA(port_ioaddr));
			outb(0x00, DATA(port_ioaddr));
			usleep(100000);
			if (inb(STAT(port_ioaddr)) != status)
				break;
		}

	printf("Warming up...\n");

	while (!terminate) {
		long long loop_timeout = rdtsc() + ns2tsc(1000000000LL);
		int inner_loops = 0;

		loop_avg = 0;
		while (rdtsc() < loop_timeout) {
			do_inner_loop();
			inner_loops++;
		}

		count += inner_loops;

		if (!warmup && !terminate) {
			loop_avg /= inner_loops;

			printf("%llu: %.3f / %.3f / %.3f us\n", count,
			       ((double)min_lat) / 1000.0,
			       ((double)loop_avg) / 1000.0,
			       ((double)max_lat) / 1000.0);

			avg_lat += loop_avg;
			outer_loops++;
		} else
			warmup = 0;
	}

	avg_lat /= outer_loops;
	printf("---\n%llu: %.3f / %.3f / %.3f us\n", count,
	       ((double)min_lat) / 1000.0, ((double)avg_lat) / 1000.0,
	       ((double)max_lat) / 1000.0);

	return 0;
}
